libstb/secvar/backend/edk2-compat-util.c

```diff
@@ -0,0 +1,601 @@
+		return NULL;
+
+	return &((struct efi_time *)timestamp_list)[off];
+}
+
+int update_timestamp(char *key, struct efi_time *timestamp)
+{
+	struct efi_time *prev;
+
+	prev = get_last_timestamp(key);
+	if (prev == NULL)
+		return OPAL_INTERNAL_ERROR;
+
+	memcpy(prev, timestamp, sizeof(struct efi_time));
+
+	prlog(PR_DEBUG, "updated prev year is %d month %d day %d\n",
+	       le16_to_cpu(prev->year), prev->month, prev->day);
+
+	return OPAL_SUCCESS;
+}
+
+int check_timestamp(char *key, struct efi_time *timestamp)
+{
+	struct efi_time *prev;
+
+	prev = get_last_timestamp(key);
+	if (prev == NULL)
+		return OPAL_INTERNAL_ERROR;
+
+	printf("timestamp year is %d month %d day %d\n", le16_to_cpu(timestamp->year), timestamp->month, timestamp->day);
+	printf("prev year is %d month %d day %d\n", le16_to_cpu(prev->year), prev->month, prev->day);
+	if (le16_to_cpu(timestamp->year) > le16_to_cpu(prev->year))
+		return OPAL_SUCCESS;
+	if (le16_to_cpu(timestamp->year) < le16_to_cpu(prev->year))
+		return OPAL_PERMISSION;
+
+	if (timestamp->month > prev->month)
+		return OPAL_SUCCESS;
+	if (timestamp->month < prev->month)
+		return OPAL_PERMISSION;
+
+	if (timestamp->day > prev->day)
+		return OPAL_SUCCESS;
+	if (timestamp->day < prev->day)
+		return OPAL_PERMISSION;
+
+	if (timestamp->hour > prev->hour)
+		return OPAL_SUCCESS;
+	if (timestamp->hour < prev->hour)
+		return OPAL_PERMISSION;
+
+	if (timestamp->minute > prev->minute)
+		return OPAL_SUCCESS;
+	if (timestamp->minute < prev->minute)
+		return OPAL_PERMISSION;
+
+	if (timestamp->second > prev->second)
+		return OPAL_SUCCESS;
+
+	/* Time less than or equal to is considered as replay*/
+	if (timestamp->second <= prev->second)
+		return OPAL_PERMISSION;
+
+	/* nanosecond, timezone, daylight and pad2 are meant to be zero */
+
+	return OPAL_SUCCESS;
+}
+
+int get_pkcs7(struct efi_variable_authentication_2 *auth, mbedtls_pkcs7 **pkcs7)
+{
+	char *checkpkcs7cert = NULL;
+	int len;
+	int rc;
+
+	len = get_pkcs7_len(auth);
+	if (len <= 0)
+		return OPAL_PARAMETER;		
+
+	if (!pkcs7)
+		return OPAL_PARAMETER;
+
+	*pkcs7 = malloc(sizeof(struct mbedtls_pkcs7));
+	if (!(*pkcs7))
+		return OPAL_NO_MEM;
+
+	mbedtls_pkcs7_init(*pkcs7);
+	rc = mbedtls_pkcs7_parse_der(
+			(const unsigned char *)auth->auth_info.cert_data,
+			(const unsigned int)len, *pkcs7);
+	if (rc) {
+		prlog(PR_ERR, "Parsing pkcs7 failed %04x\n", rc);
+		mbedtls_pkcs7_free(*pkcs7);
+		return rc;
+	}
+
+	checkpkcs7cert = zalloc(2048);
+	if (!checkpkcs7cert) {
+		mbedtls_pkcs7_free(*pkcs7);
+		return OPAL_NO_MEM;
+	}
+
+	rc = mbedtls_x509_crt_info(checkpkcs7cert, 2048, "CRT:", &((*pkcs7)->signed_data.certs));	
+	if (rc)
+		rc = OPAL_PARAMETER;
+	else
+		prlog(PR_DEBUG, "%s \n", checkpkcs7cert);
+
+	free(checkpkcs7cert);
+	mbedtls_pkcs7_free(*pkcs7);
+
+	return OPAL_SUCCESS;
+}
+
+int verify_signature(void *auth_buffer, char *newcert, uint64_t new_data_size,
+		     struct secvar *avar)
+{
+	struct efi_variable_authentication_2 *auth = NULL;
```

Comment:
Why not make this the parameter to the function, instead of immediately casting it from a `void *`?